### PR TITLE
ORC-1489: Assign a writer id to CUDF

### PR DESCRIPTION
**Pull request by guiyanakuang**
### What changes were proposed in this pull request?

This pr is aimed at assigning a writer id to the CUDF.

### Why are the changes needed?

This helps to locate the writer of a specific orc file, and it also helps the reader to do some special reads for files created by different writers.

### How was this patch tested?

Added UT

### CODE REVIEW

**Comment by wgtmac** [comment]
I assume there is a follow-up action to adopt this in the CUDF community?

**Comment by guiyanakuang** [comment]
I'm going to make a pr for CUDF later to set the id. You've reminded me not to merge this pr until it's accepted by the CUDF community.

**Comment by wgtmac** [comment]
> I'm going to make a pr for CUDF later to set the id. You've reminded me not to merge this pr until it's accepted by the CUDF community.

BTW, they cannot set this value until it is released.

**Comment by dongjoon-hyun** [comment]
Ya, +1 for @wgtmac for holding on this until we have the official release binary.

**Review by dongjoon-hyun — COMMENTED** [review]
I was a little confused about this comment. Do you mean CUDF release or Apache ORC release with ORC-1489, @wgtmac ?
> BTW, they cannot set this value until it is released.

**Comment by dongjoon-hyun** [comment]
One more comment to @guiyanakuang . I'd like to recommend to have an umbrella JIRA for all CUDA works, @guiyanakuang . :)

**Comment by omalley** [comment]
You also need to update the specification at https://github.com/apache/orc/blob/873d00c1f42753e9522cc488db59b1c979299eea/site/specification/ORCv1.md?plain=1#L138.

This PR should be merged first since it effectively reserves the number for them.

**Comment by wgtmac** [comment]
> I was a little confused about this comment. Do you mean CUDF release or Apache ORC release with [ORC-1489](https://issues.apache.org/jira/browse/ORC-1489), @wgtmac ?
> 
> > BTW, they cannot set this value until it is released.

Sorry for the confusion. I mean we need to make an official release to let cudf set the value.

**Comment by dongjoon-hyun** [comment]
Thanks, @wgtmac . Then, let's make it happen at 1.9.2 to help them because 2.0.0 is scheduled in 2024.

**Comment by guiyanakuang** [comment]
> One more comment to @guiyanakuang . I'd like to recommend to have an umbrella JIRA for all CUDA works, @guiyanakuang . :)

I have created an umbrella [JIRA](https://issues.apache.org/jira/browse/ORC-1494) for all CUDF work. Please let me know if any tasks are missing.

**Review by wgtmac — APPROVED** [review]
+1

File: site/specification/ORCv1.md [hunk 1]
```diff
@@ -136,6 +136,7 @@ message Footer {
  // 2 = Presto
  // 3 = Scritchley Go from https://github.com/scritchley/orc
  // 4 = Trino
+ // 5 = CUDF
```
Comment (wgtmac) [inline site/specification/ORCv1.md:139]: Should we update ORCv2.md as well?

**Review by dongjoon-hyun — APPROVED** [review]
I also agree with @wgtmac 's comment about `ORCv2.md`.

+1, LGTM except that.

File: site/specification/ORCv1.md (hunk 1 above)
Comment (guiyanakuang) [reply site/specification/ORCv1.md:139]: Good catch!

**Comment by dongjoon-hyun** [comment]
Merged to main/1.9/1.8. Thank you, @guiyanakuang and @wgtmac .

BTW, to @guiyanakuang , I used `dev/merge_orc_pr.py` to merge to `main/1.9/1.8` branches and update JIRA together.

**Comment by dongjoon-hyun** [comment]
Additionally, GitHub Milestone feature doesn't allow multiple versions on a single PR. So, I created a clone of this for milestone 1.9.2 separately. It's #1603.

**Comment by wgtmac** [comment]
Thanks @dongjoon-hyun! I will start the work of v1.8.5 release preparation this weekend.

**Comment by dongjoon-hyun** [comment]
@wgtmac Thank you so much always! Yes, it will be a meaningful release.

**Comment by dongjoon-hyun** [comment]
Apache ORC website is updated first.
- https://orc.apache.org/specification/ORCv1/

Screenshot 2023-08-30 at 9 35 00 AM

**Pull request closed without merging**